OSDOCS#10927: Adding support for secrets store with GCP (#82241)
openshift/openshift-docs: modules/secrets-store-google.adoc

// Module included in the following assemblies:
//
// * nodes/pods/nodes-pods-secrets-store.adoc

:_mod-docs-content-type: PROCEDURE
[id="secrets-store-google_{context}"]
= Mounting secrets from Google Secret Manager

> You can use the {secrets-store-operator} to mount secrets from Google Secret Manager to a CSI volume in {product-title}. To mount secrets from Google Secret Manager, your cluster must be installed on {gcp-first}.

Reviewer: ```suggestion
You can use the {secrets-store-operator} to mount secrets from Google Secret Manager to a Container Storage Interface (CSI) volume in {product-title}. To mount secrets from Google Secret Manager, your cluster must be installed on {gcp-first}.
```

.Prerequisites

> * You have installed the {secrets-store-operator}. See _Installing the {secrets-store-driver}_ for instructions.

Reviewer: ```suggestion
* You installed the {secrets-store-operator}. See _Installing the {secrets-store-driver}_ for instructions.
```

> * You have configured Google Secret Manager to store the required secrets.

Reviewer: ```suggestion
* You configured Google Secret Manager to store the required secrets.
```

> * You have created a service account key named `key.json` from your Google Cloud service account.

Reviewer: ```suggestion
* You created a service account key named `key.json` from your Google Cloud service account.
```

* You have access to the cluster as a user with the `cluster-admin` role.

.Procedure

> . Install the Google Secret Manager provider:

Reviewer: ```suggestion
. Install the Google Secret Manager provider by using the following steps:
```

        - name: provider
          image: us-docker.pkg.dev/secretmanager-csi/secrets-store-csi-driver-provider-gcp/plugin@sha256:a493a78bbb4ebce5f5de15acdccc6f4d19486eae9aa4fa529bb60ac112dd6650
          securityContext:
            privileged: true
          imagePullPolicy: IfNotPresent
          resources:
            requests:
              cpu: 50m
              memory: 100Mi
            limits:
              cpu: 50m
              memory: 100Mi
          env:
            - name: TARGET_DIR
              value: "/etc/kubernetes/secrets-store-csi-providers"
          volumeMounts:
            - mountPath: "/etc/kubernetes/secrets-store-csi-providers"
              name: providervol
              mountPropagation: None
              readOnly: false
          livenessProbe:
            failureThreshold: 3
            httpGet:
              path: /live
              port: 8095
            initialDelaySeconds: 5
            timeoutSeconds: 10
            periodSeconds: 30
      volumes:
        - name: providervol
          hostPath:
            path: /etc/kubernetes/secrets-store-csi-providers
      tolerations:
        - key: kubernetes.io/arch
          operator: Equal

>           value: amd64

Reviewer: Is this the only arch possible/useful or just an example value?

Author: I believe it could be adjusted just like any of the settings in these examples

          effect: NoSchedule
      nodeSelector:
        kubernetes.io/os: linux
----

.. Grant privileged access to the `csi-secrets-store-provider-gcp` service account by running the following command:
+
[source,terminal]
----
$ oc adm policy add-scc-to-user privileged -z csi-secrets-store-provider-gcp -n openshift-cluster-csi-drivers
----

.. Create the provider resources by running the following command:
+
[source,terminal]
----
$ oc apply -f gcp-provider.yaml
----

> . Grant permission to read the Google Secret Manager secret:

Reviewer: ```suggestion
. Grant permission to read the Google Secret Manager secret by using the following steps:
```

.. Create a new project by running the following command:
+
[source,terminal]
----

> $ oc new-project my-namespace

Reviewer: is `my-namespace` user-replaceable?

----

.. Label the `my-namespace` namespace for pod security admission by running the following command:
+
[source,terminal]
----
$ oc label ns my-namespace security.openshift.io/scc.podSecurityLabelSync=false pod-security.kubernetes.io/enforce=privileged pod-security.kubernetes.io/audit=privileged pod-security.kubernetes.io/warn=privileged --overwrite
----

.. Create a service account for the pod deployment:
+
[source,terminal]
----

> $ oc create serviceaccount my-service-account --namespace=my-namespace

Reviewer: same Q `my-service-account`?
seems like this procedure is using example values that can be replaced; thinking a note about that might be warranted somewhere?

Author: Yeah this procedure uses example values. They can be replaced if needed, but they don't have to be. This is similar to what we've done in the other provider docs that are already out there, so I think I'll leave this be.

----

.. Create a generic secret from the `key.json` file by running the following command:
+
[source,terminal]
----
$ oc create secret generic secrets-store-creds -n my-namespace --from-file=key.json <1>
----
<1> You created this `key.json` file from the Google Secret Manager.

.. Apply the `secrets-store.csi.k8s.io/used=true` label to allow the provider to find this `nodePublishSecretRef` secret:
+
[source,terminal]
----
$ oc -n my-namespace label secret secrets-store-creds secrets-store.csi.k8s.io/used=true
----

> . Create a secret provider class to define your secrets store provider:

Reviewer: ```suggestion
. Create a `SecretProviderClass` custom resource (CR) to define your secrets store provider by using the following steps:
```

kind: SecretProviderClass
metadata:
  name: my-gcp-provider                        <1>
  namespace: my-namespace                      <2>
spec:
  provider: gcp                                <3>
  parameters:                                  <4>
    secrets: |
      - resourceName: "projects/my-project/secrets/testsecret1/versions/1"
        path: "testsecret1.txt"
----
<1> Specify the name for the secret provider class.
<2> Specify the namespace for the secret provider class.
<3> Specify the provider as `gcp`.
<4> Specify the provider-specific configuration parameters.

.. Create the `SecretProviderClass` object by running the following command:
+
[source,terminal]
----
$ oc create -f secret-provider-class-gcp.yaml
----

. Create a deployment to use this secret provider class:

.. Create a YAML file that defines the `Deployment` object:
+
.Example `deployment.yaml`
[source,yaml]
----
apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-gcp-deployment                              <1>
  namespace: my-namespace                              <2>
spec:
  replicas: 1
  selector:
    matchLabels:
      app: my-storage
  template:
    metadata:
      labels:

>         app: my-storage

Reviewer: Do we need to specify a name for the `my-storage` value?

Author: I'm not sure what you mean - this is the value for the labels.app field. It's reference just within this same YAML block

    spec:
      serviceAccountName: my-service-account           <3>
      containers:
      - name: busybox
        image: k8s.gcr.io/e2e-test-images/busybox:1.29
        command:
          - "/bin/sleep"
          - "10000"
        volumeMounts:
        - name: secrets-store-inline
          mountPath: "/mnt/secrets-store"
          readOnly: true
      volumes:
        - name: secrets-store-inline
          csi:
            driver: secrets-store.csi.k8s.io
            readOnly: true
            volumeAttributes:
              secretProviderClass: "my-gcp-provider"   <4>
            nodePublishSecretRef:
              name: secrets-store-creds                <5>
----
<1> Specify the name for the deployment.
<2> Specify the namespace for the deployment. This must be the same namespace as the secret provider class.
<3> Specify the service account you created.
<4> Specify the name of the secret provider class.
<5> Specify the name of the Kubernetes secret that contains the service principal credentials to access Google Secret Manager.

.. Create the `Deployment` object by running the following command:
+
[source,terminal]
----
$ oc create -f deployment.yaml
----

.Verification

> * Verify that you can access the secrets from Google Secret Manager in the pod volume mount:

Reviewer: ```suggestion
* Verify that you can access the secrets from Google Secret Manager in the pod volume mount by using the following steps:
```

Author: I don't think this is required on the intro to substeps, so I'll leave this be

> .. List the secrets in the pod mount:

Reviewer: ```suggestion
.. List the secrets in the pod mount by running the following command:
```

+
[source,terminal]
----
$ oc exec busybox-<hash> -n my-namespace -- ls /mnt/secrets-store/
----
+
.Example output
[source,terminal]
----
testsecret1
----

> .. View a secret in the pod mount:

Reviewer: ```suggestion
.. View a secret in the pod mount by running the following command:
```

+
[source,terminal]
----

> $ oc exec my-gcp-deployment-<hash> -n my-namespace -- cat /mnt/secrets-store/testSecret1

Reviewer 2: ```suggestion
$ oc exec my-gcp-deployment-<hash> -n my-namespace -- cat /mnt/secrets-store/testsecret1
```
this is case sensitive and should match the output of `ls /mnt/secrets-store/`

Author: Thanks! Updated

// Module included in the following assemblies:
//
// * nodes/pods/nodes-pods-secrets-store.adoc

:_mod-docs-content-type: PROCEDURE
[id="secrets-store-google_{context}"]
= Mounting secrets from Google Secret Manager

You can use the {secrets-store-operator} to mount secrets from Google Secret Manager to a CSI volume in {product-title}. To mount secrets from Google Secret Manager, your cluster must be installed on {gcp-first}.

> // TODO: Confirm must be on GCP

Author: @ropatil010 Is this a true statement above that their cluster has to be installed on GCP?

Reviewer 3: Hi @bergerhoffer TBH the cases work on other platform like(AWS) as well. As we followed for previous things like Azure keyvault with Azure platform, AWS secrets manager with AWS so i suggested in the same way like GCP secrets manager with GCP platform. Need to get inputs from Anjali on this.

Author: Thanks I'll check with Anjali.

              value: "/etc/kubernetes/secrets-store-csi-providers"
          volumeMounts:
            - mountPath: "/etc/kubernetes/secrets-store-csi-providers"
              name: providervol
              mountPropagation: None
              readOnly: false
          livenessProbe:
            failureThreshold: 3
            httpGet:
              path: /live
              port: 8095
            initialDelaySeconds: 5
            timeoutSeconds: 10
            periodSeconds: 30
      volumes:
        - name: providervol
          hostPath:
            path: /etc/kubernetes/secrets-store-csi-providers
      tolerations:
        - key: kubernetes.io/arch
          operator: Equal
          value: amd64
          effect: NoSchedule
      nodeSelector:
        kubernetes.io/os: linux
----

.. Grant privileged access to the `csi-secrets-store-provider-gcp` service account by running the following command:
+
[source,terminal]
----
$ oc adm policy add-scc-to-user privileged -z csi-secrets-store-provider-gcp -n openshift-cluster-csi-drivers
----

.. Create the provider resources by running the following command:
+
[source,terminal]
----
$ oc apply -f gcp-provider.yaml
----
+

> // TODO: Shouldn't this step to apply come before the step to grant privileged access? But AWS and Azure do it in this order...

Reviewer 3: Hi @bergerhoffer, It works fine with both ways. 
Install GCP provider then add policy OR Add policy first then install GCP provider. 
We can follow same template similar to AWS, Azure like installing provider first then add the policy.

> // TODO: Shouldn't this step to apply come before the step to grant privileged access? But AWS and Azure do it in this order...

Author: Great thanks, I'll leave it as is then

. Grant permission to read the Google Secret Manager secret:

.. Create a new project by running the following command:
+
[source,terminal]
----
$ oc new-project my-namespace
----

.. Label the `my-namespace` namespace for pod security admission by running the following command:
+
[source,terminal]
----
$ oc label ns my-namespace security.openshift.io/scc.podSecurityLabelSync=false pod-security.kubernetes.io/enforce=privileged pod-security.kubernetes.io/audit=privileged pod-security.kubernetes.io/warn=privileged --overwrite
----

.. Create a service account:
+

> // TODO: why?

Reviewer 3: Create a Kubernetes ServiceAccount for the application(pod) to use else it wont reach to running state.

  template:
    metadata:
      labels:
        app: my-storage
    spec:
      serviceAccountName: my-service-account           <3>
      containers:
      - name: busybox
        image: k8s.gcr.io/e2e-test-images/busybox:1.29
        command:
          - "/bin/sleep"
          - "10000"
        volumeMounts:
        - name: secrets-store-inline
          mountPath: "/mnt/secrets-store"
          readOnly: true
      volumes:
        - name: secrets-store-inline
          csi:
            driver: secrets-store.csi.k8s.io
            readOnly: true
            volumeAttributes:
              secretProviderClass: "my-gcp-provider"   <4>
            nodePublishSecretRef:
              name: secrets-store-creds                <5>
----
<1> Specify the name for the deployment.
<2> Specify the namespace for the deployment. This must be the same namespace as the secret provider class.
<3> Specify the service account you created.
<4> Specify the name of the secret provider class.
<5> Specify the name of the Kubernetes secret that contains the service principal credentials to access Google Secret Manager.
+
////
- Changed mypodserviceaccount -> my-service-account
////
+

> // TODO: Confirm nodePublishSecretRef is required

Author: @ropatil010 Just double checking that nodePublishSecretRef is indeed required for GCP?

Reviewer 3: Yes, we are doing using nodePublishSecretRef method by creating generic secret using key.json, adding label

.. Create the `Deployment` object by running the following command:
+
[source,terminal]
----
$ oc create -f deployment.yaml
----

.Verification

* Verify that you can access the secrets from Google Secret Manager in the pod volume mount:

.. List the secrets in the pod mount:
+
[source,terminal]
----

> $ oc exec busybox-<hash> -n my-namespace -- ls /mnt/secrets-store/

Reviewer 3: Plz update to oc exec my-gcp-deployment-xxxx -n my-namespace -- ls /mnt/secrets-store/

Author: @ropatil010 Just double checking that this is correct? With my-gcp-deployment being the name of the deployment, but the container is named busybox. Looking at the existing AWS examples, we have busybox listed in these steps.

Reviewer 3: Hi Sorry, then better we need to update the doc in those steps as well i guess. 
It always takes podname/depname not container name actually. 
ex: 
dep.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: mydep-fs
    spec:
      containers:
      - name: hello-storage

oc exec hello-storage -n testropatil -- ls /mnt/storage 
Error from server (NotFound): pods "hello-storage" not found

oc exec mydep-fs-686596d4f7-wwrgj -n testropatil -- date            
Thu Sep 26 14:07:22 UTC 2024

@dobsonj Can you plz correct here if i'm wrong?

Reviewer 2: @ropatil010 is right, `oc exec` should reference the pod name, not the container name. So it should be `my-gcp-deployment-<hash>` instead of `busybox-<hash>`. Same goes for the other providers, it should match the pod name.
